src/audio/pipeline/pipeline-graph.c

```diff
@@ -422,48 +422,87 @@ int pipeline_for_each_comp(struct comp_dev *current,
 	return 0;
 }
 
-/* visit connected  pipeline to find the dai comp and latency */
-struct comp_dev *pipeline_get_dai_comp(uint32_t pipeline_id, uint32_t *latency)
+/* visit connected pipeline to find the dai comp */
+struct comp_dev *pipeline_get_dai_comp(uint32_t pipeline_id)
 {
 	struct ipc_comp_dev *sink;
 	struct ipc *ipc = ipc_get();
 
-	*latency = 0;
-
 	sink = ipc_get_ppl_sink_comp(ipc, pipeline_id);
-	if (!sink)
-		return NULL;
-
 	while (sink) {
-		struct comp_dev *buff_comp;
 		struct comp_buffer *buffer;
+		struct comp_dev *comp;
 
-		*latency += sink->cd->pipeline->period;
 		/* dai is found */
-		if (list_is_empty(&sink->cd->bsink_list)) {
-			struct list_item *list;
-
-			list = comp_buffer_list(sink->cd, PPL_DIR_UPSTREAM);
-			buffer = buffer_from_list(list->next, struct comp_buffer, PPL_DIR_UPSTREAM);
-			break;
-		}
+		if (list_is_empty(&sink->cd->bsink_list))
+			return sink->cd;
 
 		buffer = buffer_from_list(comp_buffer_list(sink->cd, PPL_DIR_DOWNSTREAM)->next,
 					  struct comp_buffer, PPL_DIR_DOWNSTREAM);
-		buff_comp = buffer_get_comp(buffer, PPL_DIR_DOWNSTREAM);
+		comp = buffer_get_comp(buffer, PPL_DIR_DOWNSTREAM);
 
 		/* buffer_comp is in another pipeline and it is not complete */
-		if (!buff_comp->pipeline)
+		if (!comp->pipeline)
 			return NULL;
 
-		sink = ipc_get_ppl_sink_comp(ipc, buff_comp->pipeline->pipeline_id);
+		sink = ipc_get_ppl_sink_comp(ipc, comp->pipeline->pipeline_id);
+	}
+
+	return NULL;
+}
+
+#if CONFIG_IPC_MAJOR_4
+/* visit connected pipeline to find the dai comp and latency */
```

Comment:
this function could do with a larger description explaining the algo used to calculate latency.